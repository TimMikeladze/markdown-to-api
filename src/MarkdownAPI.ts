import { globbySync } from 'globby';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch, { Options as MiniSearchOptions } from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import { parse } from 'yaml';
import deepmerge from 'deepmerge';

const toSlug = (string: string): string => slugify(string, { lower: true, strict: true });

export interface FileMetadata {
  description?: string;
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  metadata: FileMetadata;
  strippedContent: string;
}

export interface Options {
  config?: any;
  configPath?: string;
  directory: string;
  indexPath?: string;
  miniSearchOptions?: MiniSearchOptions<any>;
  useIndex?: boolean;
  writeIndex?: boolean;
}

export interface Config {
  fields?: Record<string, {
    required?: boolean;
  } | null>,
  tags?: Record<string, {
    description?: string;
  } | null>
}

const defaultConfig: Config = {
  tags: {},
  fields: {
    title: {
      required: false,
    },
    description: {
      required: false,
    },
    tags: {
      required: false,
    },
  },
};

export class MarkdownAPI {
  private static readonly stripper = remark()
    .use(strip);

  private filePaths!: string[];

  private miniSearch!: MiniSearch;

  private files!: Map<string, ParsedFile>;

  private readonly options: Options;

  private config!: Config;

  constructor(options: Options) {
    this.options = options;
    this.init();
  }

  public getOptions(): Options {
    return this.options;
  }

  public init() {
    this.config = this.loadConfig();
    this.filePaths = this.loadFilePaths();
    this.files = this.loadFiles();
    if (this.options.useIndex && existsSync(this.getIndexPath())) {
      this.miniSearch = this.loadIndexJSON();
    } else {
      this.miniSearch = this.buildIndex();
    }
    if (this.options.writeIndex) {
      this.writeIndexJSON();
    }
  }

  public parseFile(path: string): ParsedFile {
    const output = matter(readFileSync(path, 'utf8'));

    let {
      data: {
        // eslint-disable-next-line prefer-const
        id, title, slug, tags, description,
      },
    }: {
      data: Partial<FileMetadata>;
    } = output;

    if (!title || !title.trim().length) {
      title = path.split('/').pop()?.split('.').shift();
    }

    if (!slug || !slug.trim().length) {
      slug = toSlug(title as string);
    }

    if (!id) {
      const parts = path.split('/');
      id = `${parts.slice(1, parts.length - 1).join('-')}-${slug}`;
    }

    id = toSlug(id);

    if (!description) {
      description = '';
    }

    const defaultFieldKeys = Object.keys(defaultConfig.fields || {});

    const extraFieldKeys = Object.keys(this.config.fields || {})
      .map((x) => (this.config.fields?.[x] && !defaultFieldKeys.includes(x) ? x : null))
      .filter(Boolean) as string[];

    const extraFields = extraFieldKeys.reduce<Record<string, string>>((res, key) => ({
      ...res,
      [key]: (output.data as any)[key],
    }), {});

    const strippedContent = (MarkdownAPI.stripper.processSync(output.content)).toString();

    return {
      metadata: {
        id,
        title,
        slug,
        path,
        tags,
        description,
        ...extraFields,
      },
      content: output.content,
      strippedContent,
    };
  }

  public loadConfig() {
    let config = {};
    if (this.options.configPath) {
      config = parse(readFileSync(this.options.configPath, 'utf8'));
    } else if (this.options.config) {
      config = this.options.config;
    } else if (existsSync(`${this.options.directory}/config.yml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yml`, 'utf8'));
    } else if (existsSync(`${this.options.directory}/config.yaml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yaml`, 'utf8'));
    }

    return deepmerge(defaultConfig, config);
  }

  public getConfig(): Config {
    return this.config;
  }

  public getTags(): {
    description?: string;
    name: string;
  }[] {
    return Object.keys(this.config.tags || {}).map((x) => ({
      name: x,
      description: this.config.tags?.[x]?.description || '',
    }));
  }

  public loadFilePaths() {
    const paths = globbySync(this.options.directory, {
      expandDirectories: {
        files: ['*.md'],
      },
    });
    return paths;
  }

  public loadFiles(): Map<string, ParsedFile> {
    const parsedFiles = this.filePaths.map((x) => this.parseFile(x));

    const requiredFields = Object.keys(this.config.fields || {})
      .map((x) => (this.config.fields?.[x]?.required ? x : null))
      .filter(Boolean);

    let errors: string[] = [];

    const files = parsedFiles.reduce((map, file) => {
      errors = [
        ...errors,
        ...requiredFields.reduce<string[]>((res, requiredField) => {
          if (requiredField && !((file.metadata as any)[requiredField])) {
            return [...res, `Missing required field '${requiredField}' in ${file.metadata.path}`];
          }
          return res;
        }, []),
      ];

      map.set(file.metadata.id, file);

      return map;
    }, new Map<string, ParsedFile>());

    if (errors.length) {
      throw new Error(errors.join('\n'));
    }

    return files;
  }

  // eslint-disable-next-line class-methods-use-this
  public getIndexFields() {
    const fields = ['title', 'slug', 'path', 'tags', 'description'];
    return {
      fields: [...fields, 'strippedContent'],
      storeFields: fields,
    };
  }

  public buildIndex(): MiniSearch {
    const miniSearch = new MiniSearch({
      ...this.getIndexFields(),
      ...this.options.miniSearchOptions,
    });

    miniSearch.addAll(Array.from(this.files.values()).map((x) => ({ ...x.metadata, ...x })));

    return miniSearch;
  }

  public getFile(id: string): ParsedFile | undefined {
    return this.files.get(id);
  }

  public getFiles(): Map<string, ParsedFile> {
    return this.files;
  }

  public getIndex() {
    return this.miniSearch;
  }

  public indexToJSON(): any {
    return JSON.stringify(this.miniSearch.toJSON());
  }

  public writeIndexJSON(): void {
    writeFileSync(this.getIndexPath(), this.indexToJSON(), 'utf8');
  }

  public getIndexPath(): string {
    return this.options.indexPath || `${this.options.directory}/index.json`;
  }

  public loadIndexJSON() {
    if (this.getIndexPath()) {
      return MiniSearch.loadJSON(readFileSync(this.getIndexPath(), 'utf8'), {
        ...this.getIndexFields(),
        ...this.options.miniSearchOptions,
      });
    }
    throw new Error('No index path provided');
  }

  public getFilePaths(): string[] {
    return this.filePaths;
  }
}
