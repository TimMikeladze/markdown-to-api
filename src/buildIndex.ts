import { globby } from 'globby';
import { readFile } from 'fs/promises';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';

export const loadFilePaths = async (path: string) => {
  const paths = await globby(path, {
    expandDirectories: {
      files: ['*.md'],
    },
  });
  return paths;
};

export interface FileData {
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  data: FileData;
  rawContent: string;
}

const stripper = remark()
  .use(strip);

export const parseFile = async (path: string): Promise<ParsedFile> => {
  const output = matter(await readFile(path, 'utf8'));

  let {
    data: {
      // eslint-disable-next-line prefer-const
      id, title, slug, tags,
    },
  }: {
    data: Partial<FileData>;
  } = output;

  if (!title || !title.trim().length) {
    title = path.split('/').pop()?.split('.').shift();
  }

  if (!slug || !slug.trim().length) {
    slug = slugify(title as string, {
      lower: true,
      strict: true,
    });
  }

  if (!id) {
    const parts = path.split('/');
    id = `${parts.slice(0, parts.length - 1).join('/')}/${slug}`;
  }

  const rawContent = (await stripper.process(output.content)).toString();

  return {
    data: {
      id,
      title,
      slug,
      path,
      tags,
    },
    content: output.content,
    rawContent,
  };
};

export const parseFiles = (paths: string[]): Promise<ParsedFile[]> => {
  const files = Promise.all(paths.map(parseFile));
  return files;
};

export const buildIndex = async (files: ParsedFile[]) => {
  const fields = ['title', 'slug', 'path', 'tags'];

  const miniSearch = new MiniSearch({
    fields: [...fields, 'content'],
    storeFields: fields,
  });

  await miniSearch.addAllAsync(files.map((x) => ({ ...x.data, content: x.content })));

  return miniSearch;
};
