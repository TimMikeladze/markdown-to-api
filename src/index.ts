import { globby } from 'globby';
import { readFile } from 'fs/promises';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';
import { existsSync, readFileSync } from 'fs';
import { parse } from 'yaml';
import deepmerge from 'deepmerge';

export interface FileData {
  description?: string;
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  data: FileData;
  strippedContent: string;
}

export interface Options {
  config?: any;
  configPath?: string;
  directory: string;
}

export interface Config {
  fields?: Record<string, {
    required?: boolean;
  } | null>,
  tags?: Record<string, {
    description?: string;
  } | null>
}

const defaultConfig: Config = {
  tags: {},
  fields: {
    title: {
      required: false,
    },
    description: {
      required: false,
    },
    tags: {
      required: false,
    },
  },
};

export class MarkdownToGraphQL {
  private static readonly stripper = remark()
    .use(strip);

  private filePaths!: string[];

  private miniSearch!: MiniSearch;

  private files!: Map<string, ParsedFile>;

  private readonly options: Options;

  private readonly config: Config;

  constructor(options: Options) {
    this.options = options;
    this.config = this.loadConfig();
  }

  public async init() {
    this.filePaths = await this.loadFilePaths();
    this.files = await this.loadFiles();
    this.miniSearch = await this.buildIndex();
  }

  public loadConfig() {
    let config = {};
    if (this.options.configPath) {
      config = parse(readFileSync(this.options.configPath, 'utf8'));
    } else if (this.options.config) {
      config = this.options.config;
    } else if (existsSync(`${this.options.directory}/config.yml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yml`, 'utf8'));
    } else if (existsSync(`${this.options.directory}/config.yaml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yaml`, 'utf8'));
    }

    return deepmerge(defaultConfig, config);
  }

  public getConfig(): Config {
    return this.config;
  }

  public static async parseFile(path: string): Promise<ParsedFile> {
    const output = matter(await readFile(path, 'utf8'));

    let {
      data: {
        // eslint-disable-next-line prefer-const
        id, title, slug, tags, description,
      },
    }: {
      data: Partial<FileData>;
    } = output;

    if (!title || !title.trim().length) {
      title = path.split('/').pop()?.split('.').shift();
    }

    if (!slug || !slug.trim().length) {
      slug = slugify(title as string, {
        lower: true,
        strict: true,
      });
    }

    if (!id) {
      const parts = path.split('/');
      id = `${parts.slice(0, parts.length - 1).join('/')}/${slug}`;
    }

    const strippedContent = (await MarkdownToGraphQL.stripper.process(output.content)).toString();

    return {
      data: {
        id,
        title,
        slug,
        path,
        tags,
        description,
      },
      content: output.content,
      strippedContent,
    };
  }

  public async loadFilePaths() {
    const paths = await globby(this.options.directory, {
      expandDirectories: {
        files: ['*.md'],
      },
    });
    return paths;
  }

  public async loadFiles(): Promise<Map<string, ParsedFile>> {
    const parsedFiles = await Promise.all(this.filePaths.map(MarkdownToGraphQL.parseFile));

    const requiredFields = Object.keys(this.config.fields || {}).map((x) => (this.config.fields?.[x]?.required ? x : null)).filter(Boolean);

    let errors: string[] = [];

    const files = parsedFiles.reduce((map, file) => {
      errors = [
        ...errors,
        ...requiredFields.reduce<string[]>((res, requiredField) => {
          if (requiredField && !((file.data as any)[requiredField])) {
            return [...res, `Missing required field '${requiredField}' in ${file.data.path}`];
          }
          return res;
        }, []),
      ];

      map.set(file.data.id, file);

      return map;
    }, new Map<string, ParsedFile>());

    if (errors.length) {
      throw new Error(errors.join('\n'));
    }

    return files;
  }

  public async buildIndex(): Promise<MiniSearch> {
    const fields = ['title', 'slug', 'path', 'tags', 'description'];

    const miniSearch = new MiniSearch({
      fields: [...fields, 'strippedContent'],
      storeFields: fields,
    });

    await miniSearch.addAllAsync(Array.from(this.files.values()).map((x) => ({ ...x.data, content: x.content })));

    return miniSearch;
  }

  public getFile(id: string): ParsedFile | undefined {
    return this.files.get(id);
  }

  public getFiles(): Map<string, ParsedFile> {
    return this.files;
  }

  public getIndex() {
    return this.miniSearch;
  }

  public getFilePaths(): string[] {
    return this.filePaths;
  }
}
