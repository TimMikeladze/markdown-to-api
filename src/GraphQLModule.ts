import { createModule, gql } from 'graphql-modules';
import { SearchOptions } from 'minisearch';
import { MarkdownAPI, Options } from './MarkdownAPI';

export const typeDefs = gql`
  type Query {
    markdownFile(id: ID!): MarkdownFile!
    searchMarkdownFiles(text: String! options: MarkdownFileSearchOptions): MarkdownFileSearchResults!
  }

  type MarkdownFile {
    id: ID!
    path: String!
    slug: String!
    tags: [String!]
    title: String!
    description: String
    content: String!
    strippedContent: String!
  }

  type MarkdownFileSearchResults {
    count: Int!
    results: [MarkdownFileSearchResult!]!
  }

  type MarkdownFileSearchResult {
    id: ID!
    description: String
    score: Float!
    terms: [String!]
    title: String!
    slug: String!
    path: String!
    tags: [String!]
    markdownFile: MarkdownFile!
  }

  input MarkdownFileSearchOptions {
    fields: [String!]
    weights: MarkdownFileSearchOptionsWeights
    prefix: Boolean
    fuzzy: Boolean
    maxFuzzy: Int
    combineWith: MarkdownFileSearchOptionsCombineWith
  }

  enum MarkdownFileSearchOptionsCombineWith {
    AND
    OR
  }

  input MarkdownFileSearchOptionsWeights {
    fuzzy: Float!
    exact: Float!
  }
`;

export const getResolvers = (mdapi: MarkdownAPI) => ({
  Query: {
    markdownFile(_: any, { id }: { id: string }) {
      const file = mdapi.getFile(id);
      if (!file) {
        throw new Error(`File not found: ${id}`);
      }
      return {
        ...file,
        ...file.metadata,
      };
    },
    searchMarkdownFiles(_: any, { text, options }: { options: SearchOptions, text: string }) {
      const results = mdapi.getIndex().search(text, {
        ...mdapi.getOptions(),
        ...options,
      });
      return {
        count: results.length,
        results,
      };
    },
  },
  MarkdownFileSearchResult: {
    markdownFile({ id }: { id: string }) {
      if (!id) {
        throw new Error('selection set is missing id');
      }
      const file = mdapi.getFile(id);
      return {
        ...file,
        ...file!.metadata,
      };
    },
  },
});

export const createMarkdownAPIModule = (options: Options) => {
  const mdapi = new MarkdownAPI(options);

  return createModule({
    id: 'MarkdownAPIModule',
    typeDefs,
    resolvers: getResolvers(mdapi),
  });
};
