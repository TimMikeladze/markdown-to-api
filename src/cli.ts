import { Command } from 'commander';
import { MarkdownAPI } from './MarkdownAPI';

const program = new Command();

program
  .name('markdown-to-api')
  .description('CLI to interact with markdown-to-api');

program.requiredOption(
  '-d --directory <folder>',
  'Path to the folder containing the markdown files',
);

program.option(
  '-c --configPath <configPath>',
  'Path to the config file',
);

program.option(
  '-i --indexPath <indexPath>',
  'Path to the index file',
);

try {
  const res = program.parse(process.argv);

  const mdapi = new MarkdownAPI(res.opts());

  mdapi.writeIndexJSON();

  // eslint-disable-next-line no-console
  console.log(`Search index written to: ${mdapi.getIndexPath()}`);
} catch (err: any) {
  // eslint-disable-next-line no-console
  console.error(err.message);
}
