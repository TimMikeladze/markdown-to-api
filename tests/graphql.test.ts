import { print } from 'graphql';
import { createMarkdownAPIModule, MarkdownAPI, getResolvers } from '../src';

describe('createMarkdownAPIModule', () => {
  it('creates a GraphQLModule', () => {
    const module = createMarkdownAPIModule({
      directory: 'tests/cats',
    });

    expect(print(module.typeDefs as any)).toMatchSnapshot();
  });

  it('can query for a file', async () => {
    const mdapi = new MarkdownAPI({
      directory: 'tests/cats',
    });
    const file = getResolvers(mdapi).Query.markdownFile({}, { id: 'pumas-cougar' });
    expect(file).toMatchSnapshot();
  });

  it('throws error if file not found', async () => {
    const mdapi = new MarkdownAPI({
      directory: 'tests/cats',
    });
    expect(() => getResolvers(mdapi).Query.markdownFile({}, { id: 'dne' })).toThrowError();
  });

  it('can search for files', async () => {
    const mdapi = new MarkdownAPI({
      directory: 'tests/cats',
    });
    const res = getResolvers(mdapi).Query.searchMarkdownFiles({}, { text: 'cat' });
    expect(res).toMatchSnapshot();
  });

  it('can autosuggest search terms', async () => {
    const mdapi = new MarkdownAPI({
      directory: 'tests/cats',
    });
    const res = getResolvers(mdapi).Query.autoSuggestMarkdownFileSearch({}, { text: 'cat tiger' });
    expect(res).toMatchSnapshot();
  });
});
