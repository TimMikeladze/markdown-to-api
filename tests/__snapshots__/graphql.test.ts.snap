// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMarkdownAPIModule can autosuggest search terms 1`] = `
Object {
  "count": 1,
  "results": Array [
    Object {
      "score": 14.74233631787636,
      "suggestion": "cat tiger tigers",
      "terms": Array [
        "cat",
        "tiger",
        "tigers",
      ],
    },
  ],
}
`;

exports[`createMarkdownAPIModule can query for a file 1`] = `
Object {
  "content": "
# Cougar

The cougar (Puma concolor) is a large cat native to the Americas. Its range spans from the Canadian Yukon to the southern Andes in South America and is the most widespread of any large wild terrestrial mammal in the Western Hemisphere. It is an adaptable, generalist species, occurring in most American habitat types. Due to its wide range, it has many names, including puma, mountain lion, catamount and panther.

The cougar is the second-largest cat in the New World, after the jaguar (Panthera onca). Secretive and largely solitary by nature, the cougar is properly considered both nocturnal and crepuscular, although daytime sightings do occur. Despite its size, the cougar is more closely related to smaller felines, including the domestic cat (Felis catus) than to any species of the subfamily Pantherinae.

The cougar is an ambush predator that pursues a wide variety of prey. Primary food sources are ungulates, particularly deer, but it also hunts smaller prey such as rodents. It prefers habitats with dense underbrush and rocky areas for stalking, but also lives in open areas. The cougar is territorial and lives at low population densities. Individual home ranges depend on terrain, vegetation and abundance of prey. While large, it is not always the apex predator in its range, yielding prey it has killed to American black bears, grizzly bears and packs of wolves. It is reclusive and mostly avoids people. Fatal attacks on humans are rare, but increased in North America as more people entered cougar habitat and built farms.
",
  "createdAt": "2022-08-18T21:49:39.354Z",
  "description": "The cougar (Puma concolor) is a large cat native to the Americas.",
  "genus": "Puma",
  "id": "pumas-cougar",
  "path": "tests/cats/pumas/cougar.md",
  "slug": "cougar",
  "strippedContent": "Cougar

The cougar (Puma concolor) is a large cat native to the Americas. Its range spans from the Canadian Yukon to the southern Andes in South America and is the most widespread of any large wild terrestrial mammal in the Western Hemisphere. It is an adaptable, generalist species, occurring in most American habitat types. Due to its wide range, it has many names, including puma, mountain lion, catamount and panther.

The cougar is the second-largest cat in the New World, after the jaguar (Panthera onca). Secretive and largely solitary by nature, the cougar is properly considered both nocturnal and crepuscular, although daytime sightings do occur. Despite its size, the cougar is more closely related to smaller felines, including the domestic cat (Felis catus) than to any species of the subfamily Pantherinae.

The cougar is an ambush predator that pursues a wide variety of prey. Primary food sources are ungulates, particularly deer, but it also hunts smaller prey such as rodents. It prefers habitats with dense underbrush and rocky areas for stalking, but also lives in open areas. The cougar is territorial and lives at low population densities. Individual home ranges depend on terrain, vegetation and abundance of prey. While large, it is not always the apex predator in its range, yielding prey it has killed to American black bears, grizzly bears and packs of wolves. It is reclusive and mostly avoids people. Fatal attacks on humans are rare, but increased in North America as more people entered cougar habitat and built farms.
",
  "tags": Array [
    Object {
      "description": "Fits into any box.",
      "id": "cat",
      "name": "cat",
    },
    Object {
      "description": "Puma is a large cat, also known as a mountain lion.",
      "id": "puma",
      "name": "cougar / puma",
    },
  ],
  "title": "Cougar",
}
`;

exports[`createMarkdownAPIModule can search for files 1`] = `
Object {
  "count": 5,
  "results": Array [
    Object {
      "createdAt": "2022-08-18T21:49:39.354Z",
      "description": "The cougar (Puma concolor) is a large cat native to the Americas.",
      "id": "pumas-cougar",
      "match": Object {
        "cat": Array [
          "tags",
          "description",
          "strippedContent",
        ],
      },
      "path": "tests/cats/pumas/cougar.md",
      "score": 1.6449438295294794,
      "slug": "cougar",
      "tags": Array [
        "cat",
        "puma",
      ],
      "terms": Array [
        "cat",
      ],
      "title": "Cougar",
    },
    Object {
      "createdAt": "2022-08-18T21:49:39.355Z",
      "description": "The tiger (Panthera tigris) is the largest living cat species and a member of the genus Panthera.",
      "id": "tigers-about-tigers",
      "match": Object {
        "cat": Array [
          "tags",
          "description",
          "strippedContent",
        ],
      },
      "path": "tests/cats/tigers/about-tigers.md",
      "score": 1.4145958488973376,
      "slug": "about-tigers",
      "tags": Array [
        "cat",
        "tiger",
      ],
      "terms": Array [
        "cat",
      ],
      "title": "About Tigers",
    },
    Object {
      "createdAt": "2022-08-18T21:49:39.354Z",
      "description": "The lion (Panthera leo) is a large cat of the genus Panthera native to Africa and India. It has a muscular, broad-chested body, short, rounded head, round ears, and a hairy tuft at the end of its tail.",
      "id": "lions-about-lions",
      "match": Object {
        "cat": Array [
          "tags",
          "description",
          "strippedContent",
        ],
      },
      "path": "tests/cats/lions/about-lions.md",
      "score": 1.2603093569990829,
      "slug": "about-lions",
      "tags": Array [
        "cat",
        "lion",
      ],
      "terms": Array [
        "cat",
      ],
      "title": "About Lions",
    },
    Object {
      "createdAt": "2022-08-18T21:49:39.355Z",
      "description": "The Bengal tiger is a population of the Panthera tigris tigris subspecies. It ranks among the biggest wild cats alive today. It is considered to belong to the world's charismatic megafauna.",
      "id": "tigers-bengal-tiger",
      "match": Object {
        "cat": Array [
          "tags",
          "strippedContent",
        ],
      },
      "path": "tests/cats/tigers/bengal-tiger.md",
      "score": 0.5486338138728774,
      "slug": "bengal-tiger",
      "tags": Array [
        "cat",
        "tiger",
      ],
      "terms": Array [
        "cat",
      ],
      "title": "Bengal Tiger",
    },
    Object {
      "createdAt": "2022-08-18T21:49:39.355Z",
      "description": "The Indochinese tiger is a population of the Panthera tigris tigris subspecies that is native to Southeast Asia.",
      "id": "tigers-indochinese-tiger",
      "match": Object {
        "cat": Array [
          "tags",
        ],
      },
      "path": "tests/cats/tigers/indochinese-tiger.md",
      "score": 0.13051706548444453,
      "slug": "indochinese-tiger",
      "tags": Array [
        "cat",
        "tiger",
      ],
      "terms": Array [
        "cat",
      ],
      "title": "Indochinese Tiger",
    },
  ],
}
`;

exports[`createMarkdownAPIModule creates a GraphQLModule 1`] = `
Array [
  "type Query {
  countMarkdownFiles: Int
  markdownFile(id: ID!): MarkdownFile!
  markdownFiles(direction: MarkdownFilesSortDirection, limit: Int, offset: Int): [MarkdownFile!]!
  searchMarkdownFiles(text: String!, options: MarkdownFileSearchOptions, limit: Int, offset: Int): MarkdownFileSearchResults!
  autoSuggestMarkdownFileSearch(text: String!, options: MarkdownFileSearchOptions): MarkdownFileAutoSuggestResults!
  markdownFileTags: [MarkdownFileTag!]!
}

type MarkdownFile {
  content: String!
  strippedContent: String!
  id: ID!
  path: String!
  slug: String!
  tags: [MarkdownFileTag!]!
  title: String!
  description: String
  createdAt: String!
}

type MarkdownFileSearchResults {
  count: Int!
  results: [MarkdownFileSearchResult!]!
}

type MarkdownFileSearchResult {
  id: ID!
  description: String
  score: Float!
  terms: [String!]
  title: String!
  slug: String!
  path: String!
  tags: [MarkdownFileTag!]!
  createdAt: String!
  markdownFile: MarkdownFile!
}

type MarkdownFileTag {
  id: ID!
  name: String!
  description: String
}

input MarkdownFileSearchOptions {
  fields: [String!]
  weights: MarkdownFileSearchOptionsWeights
  prefix: Boolean
  fuzzy: Boolean
  maxFuzzy: Int
  combineWith: MarkdownFileSearchOptionsCombineWith
}

type MarkdownFileAutoSuggestResults {
  count: Int!
  results: [MarkdownFileAutoSuggestResult!]!
}

type MarkdownFileAutoSuggestResult {
  suggestion: String!
  terms: [String!]!
  score: Float!
}

enum MarkdownFileSearchOptionsCombineWith {
  AND
  OR
}

input MarkdownFileSearchOptionsWeights {
  fuzzy: Float!
  exact: Float!
}

enum MarkdownFilesSortDirection {
  asc
  desc
}

enum MarkdownFilesSortBy {
  title
  createdAt
}",
]
`;
